### ISSUE TITLE
Sometimes System.AccessViolationException is thrown when MuPDFRenderer is disposing

### ISSUE THREAD

**Issue by lexboss777** [arklumpus/MuPDFCore#35]
Hi! Recently I started looking at this library and it seems very useful.
But very often I receive an exception mentioned above. It occurs when I close the window with MuPDFRenderer. I see that renderer has callback `ControlDetachedFromLogicalTree` and invokes disposing methods in there.

```
>	MuPDFCore.dll!MuPDFCore.MuPDFContext.Dispose(bool disposing) Line 182	C#
 	MuPDFCore.dll!MuPDFCore.MuPDFContext.Dispose() Line 196	C#
 	MuPDFCore.dll!MuPDFCore.MuPDFMultiThreadedPageRenderer.Dispose(bool disposing) Line 583	C#
 	MuPDFCore.dll!MuPDFCore.MuPDFMultiThreadedPageRenderer.Dispose() Line 594	C#
 	MuPDFCore.MuPDFRenderer.dll!MuPDFCore.MuPDFRenderer.PDFRenderer.ReleaseResources() Line 625	C#
 	MuPDFCore.MuPDFRenderer.dll!MuPDFCore.MuPDFRenderer.PDFRenderer.ControlDetachedFromLogicalTree(object sender, Avalonia.LogicalTree.LogicalTreeAttachmentEventArgs e) Line 645	C#
```

I've seen that the previous issue is also about that but it's already closed and it seems the discussion were focused in disposing order when invoked manually, but I am not invoking any disposing methods on my own, it's done by MuPDFRenderer itself.

**Comment by arklumpus** [comment]
Hi, thank you for pointing this out! You're right, it's a different issue than the other one... I think I know how to solve this, but I would need to reproduce it first. Can you share more information about your configuration?
* What version of .NET are you using? .NET Framework?
* What architecture are you targeting? E.g., Any CPU, x86 or x64?
* What Avalonia version are you using?

If you have a small application example that reliably shows the issue it would be great!

**Comment by lexboss777** [comment]
I tried it with .NET 6 and .NET 8. targeting Any CPU.
I am using the latest Avalonia packages available at the moment of writing: 11.0.11.
I'll attach the repro project soon, thanks!

**Comment by lexboss777** [comment]
Please, see reproducing project here https://github.com/lexboss777/MuPDFCore_AccessViolationException
I've just took sample project from AvaloniaUI/Avalonia.Samples and modified it a bit.
The app will download sample PDF file once (the line is PdfWindow.axaml.cs::51).

To reproduce:
1. Build and run
2. Tap green round button at the bottom right corner. The pdf window will appear.
3. Close the window
4. Repeat 2-3 till the exception is thrown (it might take from 2 to 20 times)

**Comment by arklumpus** [comment]
Thank you for the repo, I think I found the issue(s):
1) The `MuPDFStructuredTextPage` held by the renderer control is not disposed
2) Context clones get disposed multiple times
3) The font pointer cache is global instead of being per-context

It should be relatively easy to fix these, but I am going away for a couple of weeks, so I won't be able to do anything until I come back... If the issue affects you too much, I would suggest downgrading to version 1.9.0 in the meantime!

**Comment by arklumpus** [comment]
Sorry for the delay! The issue should be fixed in version 1.10.2.

**TL;DR:** please update the NuGet packages and move the calls to `Document?.Dispose` and `Context?.Dispose` to the `Closed` event of the `Window`, rather than `Closing`.

The main problem was what I mentioned before, i.e. the `MuPDFStructuredTextPage` not being disposed by renderer control. As a result, its `Dispose` method was invoked by the finalizer at a random moment; depending on what else was going on with the memory management, this sometimes worked anyways, and other times it would cause the `AccessViolationException`.

However, there was also a problem with your code: you were calling `Document?.Dispose` and `Context?.Dispose` within the `Window.Closing` handler. Since this event is invoked _before_ the `DetachedFromLogicalTree` event, the context was disposed before the renderer control had the chance to release its resources.

The solution to this is to call `Document?.Dispose` and `Context?.Dispose` in the `Closed` event, which is indeed invoked _after_ `DetachedFromLogicalTree`. Alternatively, you will need to explicitly invoke `PDFRenderer.ReleaseResources` before disposing the document and context.

I hope this makes sense! In the new version, I have also made some changes to detect this issue and throw an exception with a more informative message, that gets thrown _every time_. You should see this if you update the package reference without fixing the issue I mentioned about your code.

I'm closing this for now, please feel free to reopen (or open a new issue) if you still have problems!
